File: Models/BMS/BMS_CODE/BMS.cpp
/*****************************************************************************
*	Implemetation of the saliency detction method described in paper
*	"Saliency Detection: A Boolean Map Approach", Jianming Zhang, 
*	Stan Sclaroff, ICCV, 2013
*	
*	Copyright (C) 2013 Jianming Zhang
*
*	This program is free software: you can redistribute it and/or modify
*	it under the terms of the GNU General Public License as published by
*	the Free Software Foundation, either version 3 of the License, or
*	(at your option) any later version.
*
*	This program is distributed in the hope that it will be useful,
*	but WITHOUT ANY WARRANTY; without even the implied warranty of
*	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*	GNU General Public License for more details.
*
*	You should have received a copy of the GNU General Public License
*	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*	If you have problems about this software, please contact: jmzhang@bu.edu
*******************************************************************************/

#include "BMS.h"

#include <vector>
#include <cmath>
#include <ctime>
#include <opencv2/imgproc.hpp>

using namespace cv;
using namespace std;


BMS::BMS(const Mat& src, const int dw1, const int ow, const bool nm, const bool hb)
	:_rng(),_dilation_width_1(dw1),_opening_width(ow),_normalize(nm),_handle_border(hb)
{
	_src=src.clone();
	Mat lab;
	cvtColor(_src,lab,COLOR_RGB2Lab);

	vector<Mat> maps;
	maps.push_back(lab);

	for (int i=0;i<maps.size();i++)
	{
		vector<Mat> sp;
		split(maps[i],sp);
		_feature_maps.push_back(sp[0]);
		_feature_maps.push_back(sp[1]);
		_feature_maps.push_back(sp[2]);
	}
	_sm=Mat::zeros(src.size(),CV_64FC1);
}

void BMS::computeSaliency(float step)
{
	for (int i=0;i<_feature_maps.size();++i)
	{
		double max_,min_;
		minMaxLoc(_feature_maps[i],&min_,&max_);
		for (float thresh=min_;thresh<max_;thresh+=step)
		{
			Mat bm=_feature_maps[i]>thresh;
			registerPosition(bm);
			bm=_feature_maps[i]<=thresh;
			registerPosition(bm);
		}
	}
}

Mat BMS::registerPosition(const Mat& bm)
{
	Mat bm_=bm.clone();
	if (_opening_width>0)
	{
		dilate(bm,bm_,Mat(),Point(-1,-1),_opening_width);
		erode(bm_,bm_,Mat(),Point(-1,-1),_opening_width);
	}
	
	Mat innovation=getAttentionMap(bm_);
	_sm=_sm+innovation;
	return innovation;
}


Mat BMS::getAttentionMap(const Mat& bm)
{
	Mat ret=bm.clone();
	int jump;
	if (_handle_border)
	{
		for (int i=0;i<bm.rows;i++)
		{
			jump= _rng.uniform(0.0,1.0)>0.99 ? _rng.uniform(5,25):0;
			if (ret.at<char>(i,0+jump)!=1)
				floodFill(ret,Point(0+jump,i),Scalar(1),0,Scalar(0),Scalar(0),8);
			jump = _rng.uniform(0.0,1.0)>0.99 ?_rng.uniform(5,25):0;
			if (ret.at<char>(i,bm.cols-1-jump)!=1)
				floodFill(ret,Point(bm.cols-1-jump,i),Scalar(1),0,Scalar(0),Scalar(0),8);
		}
		for (int j=0;j<bm.cols;j++)
		{
			jump= _rng.uniform(0.0,1.0)>0.99 ? _rng.uniform(5,25):0;
			if (ret.at<char>(0+jump,j)!=1)
				floodFill(ret,Point(j,0+jump),Scalar(1),0,Scalar(0),Scalar(0),8);
			jump= _rng.uniform(0.0,1.0)>0.99 ? _rng.uniform(5,25):0;
			if (ret.at<char>(bm.rows-1-jump,j)!=1)
				floodFill(ret,Point(j,bm.rows-1-jump),Scalar(1),0,Scalar(0),Scalar(0),8);
		}
	}
	else
	{
		for (int i=0;i<bm.rows;i++)
		{
			if (ret.at<char>(i,0)!=1)
				floodFill(ret,Point(0,i),Scalar(1),0,Scalar(0),Scalar(0),8);
			if (ret.at<char>(i,bm.cols-1)!=1)
				floodFill(ret,Point(bm.cols-1,i),Scalar(1),0,Scalar(0),Scalar(0),8);
		}
		for (int j=0;j<bm.cols;j++)
		{
			if (ret.at<char>(0,j)!=1)
				floodFill(ret,Point(j,0),Scalar(1),0,Scalar(0),Scalar(0),8);
			if (ret.at<char>(bm.rows-1,j)!=1)
				floodFill(ret,Point(j,bm.rows-1),Scalar(1),0,Scalar(0),Scalar(0),8);
		}
	}
	
	double max_, min_;
	minMaxLoc(ret,&min_,&max_);
	ret=ret != 1;
	
	if(_dilation_width_1>0)
		dilate(ret,ret,Mat(),Point(-1,-1),_dilation_width_1);
	ret.convertTo(ret,CV_64FC1);
	if (_normalize)
		normalize(ret,ret,1.0,0.0,NORM_L2);
	else
		normalize(ret,ret,1.0,0.0,NORM_MINMAX);
	return ret;
}

Mat BMS::getSaliencyMap()
{
	Mat ret; 
	normalize(_sm,ret,255.0,0.0,NORM_MINMAX);
	ret.convertTo(ret,CV_8UC1);
	return ret;
}

File: Models/BMS/BMS.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%	Implemetation of the saliency detction method described in paper
%	"Saliency Detection: A Boolean Map Approach", Jianming Zhang, 
%	Stan Sclaroff, ICCV, 2013
%	
%	Copyright (C) 2013 Jianming Zhang
%
%	This program is free software: you can redistribute it and/or modify
%	it under the terms of the GNU General Public License as published by
%	the Free Software Foundation, either version 3 of the License, or
%	(at your option) any later version.
%
%	This program is distributed in the hope that it will be useful,
%	but WITHOUT ANY WARRANTY; without even the implied warranty of
%	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%	GNU General Public License for more details.
%
%	You should have received a copy of the GNU General Public License
%	along with this program.  If not, see <http://www.gnu.org/licenses/>.
%
%	If you have problems about this software, please contact: jmzhang@bu.edu
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


function BMS(input_dir, output_dir, sod)
% A matlab wrapper for excuting BMS.exe. Make sure to copy the compiled
% executable file in the same directory of this wrapper. Moreover,OpenCV
% library should be in the system's path.
%
% The input and output dir should end with path sperator '/' or '\' depending
% on the operating system.
%
% **sod** is a boolean value indicating whether to use the salient object
% detection mode

if input_dir(end) ~= '/' && input_dir(end) ~= '\'
    input_dir           =   [input_dir,'/'];
end

if output_dir(end) ~= '/' && output_dir(end) ~= '\'
    output_dir          =   [output_dir,'/'];
end

if ~exist(output_dir,'dir')
    mkdir(output_dir);
end

if ~sod % for eye fixation prediction
   sample_step_size     =   8;  % \delta
   opening_width        =   5;  % \omega_o
   dilation_width_1     =   7;  % \omega_{d1}
   dilation_width_2     =   23; % \omega_{d2}
   blur_std             =   20; % \sigma
   
   use_normalization    =   1;  % L2-normalization
   handle_border        =   0;  % Some images have thin artifical black 
                                % borders surrounding them, which can 
                                % affect the output of BMS. You can 
                                % certainly trim the image before 
                                % processing it. Here we provide an adhoc 
                                % method to deal with it. When doing
                                % floodfill, we randomly put some seeds at
                                % positions near the image borders, instead
                                % of putting them exactly at the borders.
                                % We use this method in our experiments of
                                % salient object detection, because some of
                                % the testing images have this issue.
else % for salient object detection
    sample_step_size    =   8;  % \delta
    opening_width       =   13; % \omega_o
    dilation_width_1    =   1;  % \omega_{d1} (turn off dilation)
    dilation_width_2    =   1;  % \omega_{d2} (turn off dilation)
    blur_std            =   0;  % \sigma (turn off bluring)
   
    use_normalization   =   0;  % L2-normalization
    handle_border       =   1;
end

command = sprintf('./BMS %s %s %d %d %d %d %f %d %d',input_dir,output_dir,...
    sample_step_size,opening_width,dilation_width_1,dilation_width_2,...
    blur_std,use_normalization,handle_border);
system(command);

% post-processing for salient object detection tasks
if sod
    radius              =   15;
    D                   =   dir(fullfile(output_dir,'*.png'));
    file_list           =   {D.name};
    for i               =   1:numel(file_list)
        I               =   imread(fullfile(output_dir,file_list{i}));
        I               =   postProcSOD(I,radius);
        imwrite(I,fullfile(output_dir,[file_list{i}(1:end-4),'.png']));
    end
end

end

%% Post-processing function for Salient Object Detection

function sMap           =   postProcSOD(mAttMap,radius)
% opening by reconstruction followed by closing by reconstruction
% see the following material for detailed explanations 
% http://www.mathworks.com/products/demos/image/watershed/ipexwatershed.html

img_size                =   size(mAttMap);
I                       =   imresize(mAttMap,[NaN 400]); %  width = 400
se                      =   strel('disk',radius);
Ie                      =   imerode(I, se);
Iobr                    =   imreconstruct(Ie, I);
Iobrd                   =   imdilate(Iobr, se);
Iobrcbr                 =   imreconstruct(imcomplement(Iobrd), imcomplement(Iobr));
Iobrcbr                 =   imcomplement(Iobrcbr);
sMap                    =   imresize(mat2gray(Iobrcbr),img_size(1:2));

end


